ANTLR4 Grammar for ClickHouse and new parser

abyss7 opened the pull request:
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Implement new parser based on ANTLR4 runtime and generated from EBNF grammar.

abyss7 committed:
Describe SELECT statement

Generate basic fuzzers with grammarinator.

abyss7 committed:
Add generated lexer and parser on C++

abyss7 committed:
[WIP] Adopt new Parser Tree to old AST

abyss7 committed:
[WIP] waiting for better src/CMakeLists.txt

abyss7 committed:
[WIP] Skeleton of new AST

abyss7 committed:
It compiles!

abyss7 committed:
Parse and convert to old tree a simple select statement

abyss7 committed:
'SELECT 1 FROM a;' works

abyss7 committed:
[WIP]

abyss7 committed:
Convert literals to old format

abyss7 committed:
select a.b from c.d;

abyss7 committed:
Update grammar-fuzzer

abyss7 committed:
It compiles!

abyss7 committed:
Remove ColumnFunctionExpr as redundant

and merge it into ColumnExprFunction subrule.

abyss7 committed:
Temporary disable CAST function.

All fuzzed inputs are parsed successfully

abyss7 committed:
Minor fix

abyss7 committed:
Fix grammar for problems found by fuzzer

abyss7 committed:
Merge branch 'master' into antlr-grammar

abyss7 committed:
Major refactoring of ColumnExpr

convert almost everything to FUNCTION type.

abyss7 committed:
More fixes after dumping new AST

abyss7 committed:
Forbid to use NULL_SQL as identifier

abyss7 committed:
Almost full conversion from new AST to old one.

abyss7 committed:
Convert whole query list - not only the first one.

abyss7 committed:
Add support for DROP query

abyss7 committed:
Rename utils/grammar/fuzzer → utils/grammar-fuzzer

abyss7 committed:
Add support for CREATE TABLE and CREATE DATABASE queries

abyss7 committed:
Add support for complex column types in CREATE TABLE

abyss7 committed:
Add support for INSERT query with values

abyss7 committed:
Better number literal support

abyss7 committed:
Even better number literals

abyss7 committed:
Separate scalar and non-scalar, single-column and multi-column expressions

abyss7 committed:
First 545 of 1147 tests passes

melin commented:
ColumnExprTupleAccess failure
![image](https://user-images.githubusercontent.com/1145830/90794250-26b40c00-e33f-11ea-81f5-e85321af0eb3.png)

abyss7 committed:
Support more queries

abyss7 committed:
Add limited support for ALTER query

abyss7 committed:
ALTER TABLE MODIFY

abyss7 committed:
ALTER TABLE COMMENT

abyss7 committed:
CREATE MATERIALIZED VIEW

abyss7 committed:
for f in tests/queries/0_stateless/00{0,1}*.sql; do

echo "$f";
cat $f | build/utils/syntax-analyzer/syntax-analyzer 2>&1 1>/dev/null;
done

abyss7 committed:
[WIP]

abyss7 committed:
Merge branch 'master' into antlr-grammar

abyss7 committed:
Fixes after ASAN

abyss7 committed:
More fixes

abyss7 committed:
Bring back CAST function

abyss7 committed:
Fix broken tuple access

because of nasty '.number' floating point literal

abyss7 committed:
First 300 tests are parsed now

abyss7 committed:
Refactor Literal

abyss7 committed:
Minor grammar improvement

abyss7 committed:
Revert executeQuery.cpp to use old parser for now

abyss7 committed:
Loosen grammar strictness in some places

abyss7 committed:
Merge remote-tracking branch 'upstream/master' into antlr-grammar

abyss7 committed:
More tests now pass

abyss7 committed:
Tests pass 1428 of 1844

abyss7 committed:
[WIP]

abyss7 committed:
Merge branch 'master' into antlr-grammar

abyss7 committed:
Fixes for clang-11

abyss7 committed:
[WIP]

abyss7 committed:
[WIP]

abyss7 committed:
[WIP]

abyss7 committed:
Better integration of new ASTTableIdentifier

abyss7 committed:
Merge remote-tracking branch 'upstream/master' into antlr-grammar

abyss7 committed:
Untrack unused auto-generated ANTLR files

abyss7 committed:
Fix ASTIdentifier ctor invocation

abyss7 committed:
Allow to print AST for CREATE TABLE

abyss7 committed:
Merge remote-tracking branch 'upstream/master' into antlr-grammar

abyss7 committed:
Add some convertToOld() stubs

abyss7 committed:
Fix build for contrib

abyss7 committed:
[WIP]

abyss7 commented on src/Databases/DatabaseOnDisk.cpp:
```diff
@@ -101,8 +101,6 @@ String getObjectDefinitionFromCreateQuery(const ASTPtr & query)
 
     /// We remove everything that is not needed for ATTACH from the query.
     create->database.clear();
-    create->as_database.clear();
```
Not sure if it's safe to remove like this.

abyss7 committed:
Minor refactoring

abyss7 committed:
Merge remote-tracking branch 'upstream/master' into antlr-grammar

abyss7 committed:
Comment warning for clang-11

abyss7 committed:
Don't parse multiple queries and enable new parser in settings

abyss7 committed:
Migrate to Python3

abyss7 committed:
Revert ASTTableIdentifier code #1

abyss7 committed:
Revert ASTTableIdentifier code #2

abyss7 committed:
Revert ASTTableIdentifier code #3

abyss7 committed:
Revert ASTTableIdentifier #4

abyss7 committed:
Revert ASTTableIdentifier #5

abyss7 committed:
Merge remote-tracking branch 'upstream/master' into pytest

abyss7 committed:
Merge branch 'pytest' into antlr-grammar

abyss7 committed:
Revert ASTTableIdentifier #6

abyss7 committed:
Fix syntax-analyzer build

abyss7 committed:
Fix build for gcc

abyss7 committed:
Fixes for running query tests with use_antlr_parser=1

abyss7 committed:
More fixes

abyss7 committed:
Better insert support

abyss7 committed:
Add SHOW DATABASES support

abyss7 committed:
Better Nested support

abyss7 committed:
Better JOIN ON/USING support

abyss7 committed:
Better FORMAT support

abyss7 commented:
Status for now (0_stateless): `1369 failed, 456 passed, 1 skipped in 325.03s (0:05:25)`

abyss7 committed:
Support SELECT DISTINCT

abyss7 committed:
More fixes

abyss7 committed:
Refactor childrens

abyss7 committed:
Merge remote-tracking branch 'upstream/master' into antlr-grammar

abyss7 committed:
Fix build

abyss7 committed:
Refactor ColumnExpr operations precedence

abyss7 committed:
Fix typo

abyss7 committed:
Parse floating-point literals

abyss7 committed:
Attemp to fix CAST AS

abyss7 committed:
Remove ASTAlterCommandsList

abyss7 committed:
Don't allow to use keywords as aliases

abyss7 committed:
Fix tests

abyss7 committed:
More fixes

abyss7 committed:
Merge remote-tracking branch 'upstream/master' into antlr-grammar

abyss7 committed:
Fix build ALL

abyss7 committed:
More fixes

abyss7 committed:
Ignore auto-generated parser in style check

abyss7 commented:
Status for now (0_stateless): `923 failed, 982 passed in 440.90s (0:07:20)`

abyss7 committed:
More fixes

abyss7 committed:
Attempt to arcadia build

abyss7 committed:
More fixes

abyss7 committed:
Parse CONSTRAINT table element

abyss7 committed:
More fixes

abyss7 committed:
More fixes

abyss7 committed:
More fixes

abyss7 committed:
Merge remote-tracking branch 'upstream/master' into antlr-grammar

abyss7 committed:
Fix Enum type parsing

abyss7 committed:
Do not format empty args in some cases

abyss7 committed:
Do not format empty args in some cases

abyss7 commented:
Status for now (0_stateless): `581 failed, 1311 passed`

abyss7 committed:
String interpolation in literals

abyss7 committed:
More fixes

abyss7 committed:
Support LIMIT BY

abyss7 committed:
Almost support SAMPLE clause

abyss7 commented:
Status for now (0_stateless): `501 failed, 1382 passed`

abyss7 committed:
Merge remote-tracking branch 'upstream/master' into antlr-grammar

abyss7 committed:
Fix test

abyss7 committed:
More fixes

abyss7 committed:
Merge remote-tracking branch 'upstream/master' into antlr-grammar

abyss7 committed:
Fix build

abyss7 committed:
Fix tests

abyss7 committed:
Merge branch 'master' into antlr-grammar

abyss7 committed:
More fixes

abyss7 committed:
Add  support for UTF-8 queries

abyss7 committed:
Fix precedence of unary operators

abyss7 committed:
Convert SETTINGS of ENGINE properly

abyss7 committed:
Escape string literals properly

abyss7 committed:
Add support for SYSTEM query

abyss7 commented:
Status for now (0_stateless): `443 failed, 1533 passed in 698.17s (0:11:38)`

abyss7 committed:
Fix parsing of nested tuple access: t.1.2

abyss7 committed:
More fixes

abyss7 committed:
Fix gcc-10 build

abyss7 committed:
Merge remote-tracking branch 'upstream/master' into antlr-grammar

abyss7 committed:
Replace ANALYZE with EXPLAIN SYNTAX

abyss7 committed:
Fix test

abyss7 committed:
More fixes

abyss7 committed:
[WIP]

abyss7 committed:
More fixes

abyss7 committed:
More fixes

abyss7 committed:
A lot test fixes

abyss7 committed:
A lot test fixes

abyss7 committed:
Merge remote-tracking branch 'upstream/master' into antlr-grammar

abyss7 committed:
More fixes

abyss7 committed:
Merge branch 'master' into antlr-grammar

abyss7 committed:
Try to fix CI checks

abyss7 committed:
More fixes

abyss7 committed:
Merge remote-tracking branch 'upstream/master' into antlr-grammar

abyss7 committed:
Fix build

abyss7 committed:
Merge remote-tracking branch 'upstream/master' into pytest

abyss7 committed:
Merge branch 'tcp-port' into pytest

abyss7 committed:
More tests are fixed

abyss7 committed:
Merge remote-tracking branch 'upstream/master' into pytest

abyss7 committed:
Merge remote-tracking branch 'upstream/master' into antlr-grammar

abyss7 committed:
Fix ya.make

abyss7 committed:
Merge branch 'pytest' into antlr-grammar

abyss7 committed:
Try to run integration tests with new parser

abyss7 committed:
Fix style

abyss7 committed:
Fix tests

abyss7 committed:
Fix typo

abyss7 committed:
More fixes for tests

abyss7 committed:
Merge branch 'engine-format' into antlr-grammar

abyss7 committed:
More fixes

abyss7 committed:
Parse ON CLUSTER clause

abyss7 committed:
Merge branch 'master' into antlr-grammar

abyss7 committed:
More fixes

abyss7 committed:
Flatten some binary operators in new AST

abyss7 committed:
More fixes

abyss7 committed:
Fix ya.make

abyss7 committed:
More fixes

abyss7 committed:
Support Dictionary syntax

abyss7 committed:
More fixes

abyss7 committed:
Merge remote-tracking branch 'upstream/master' into antlr-grammar

abyss7 committed:
Fix build

abyss7 committed:
More fixes

abyss7 committed:
More fixes

abyss7 committed:
Fix most of the build

abyss7 committed:
Try to run functional tests with new parser

abyss7 committed:
Fix server crash

abyss7 committed:
Another segfault fix

abyss7 committed:
Run all tests with old parser

abyss7 committed:
Use $CLICKHOUSE_DATABASE in shell-tests

abyss7 committed:
Fix kill_mutation test reference

abyss7 committed:
Skip test in arcadia

abyss7 committed:
Merge branch 'master' into antlr-grammar

abyss7 committed:
Merge remote-tracking branch 'upstream/master' into antlr-grammar

baibaichen commented on src/Parsers/New/CMakeLists.txt:
```diff
@@ -0,0 +1,97 @@
+
+target_compile_options (clickhouse_parsers_new
+    PRIVATE
+        -Wno-c++20-compat
```
I use clang 11, it doesn't support this parameter, using `-Wno-c++2a-compat` instead.

```
Apple clang version 11.0.3 (clang-1103.0.32.29)
Target: x86_64-apple-darwin19.6.0
Thread model: posix
InstalledDir: /Library/Developer/CommandLineTools/usr/bin
```

abyss7 replied:
We don't support build with native Apple Clang compiler. Even if it works now, it may accidentally break in any PR since we doesn't test it in CI. It's better to use Clang from LLVM.

baibaichen commented on src/Parsers/New/AST/INode.h:
```diff
@@ -0,0 +1,103 @@
+        void dump() const { dump(0); }
+
+    protected:
+        INode() = default;
```
compile error
> definition of implicit copy constructor for 'INode' is deprecated because it has a user-declared destructor [-Werror,-Wdeprecated]

so, we need add default copy constructor

`INode(const INode&) = default;`

abyss7 replied:
Please, provide your build configuration, we don't have this kind of errors in our CI builds.

azat commented:
Introduces some `-Wsuggest-destructor-override` warnings

azat commented on tests/queries/0_stateless/01278_min_insert_block_size_rows_for_materialized_views.sh:
```diff
@@ -53,10 +53,10 @@ EOL
 echo "create table null_01278 as data_01278 Engine=Null();" | execute
 for i in $(seq 1 $TEST_01278_PARTS); do
-    echo "create table part_01278_$i as data_01278 Engine=Buffer(currentDatabase(), null_01278, 1, 86400, 86400, 1e5, 1e6, 10e6, 100e6);"
+    echo "create table part_01278_$i as data_01278 Engine=Buffer('$CLICKHOUSE_DATABASE', null_01278, 1, 86400, 86400, 1e5, 1e6, 10e6, 100e6);"
```
BTW this looks like some intermediate state? (i.e. during developing currentDatabase() was broken at some point, but fixed later?)

abyss7 replied:
I have to investigate, but for some reason the `currentDatabase()` doesn't work in some/all shell-tests, leading to very unexpected behavior: probably it's some real problems in the code which cannot properly interpret function vs literal.

abyss7 commented:
> Introduces some `-Wsuggest-destructor-override` warnings

In what build configurations have you encountered them? I added suppressions to CMake files where appropriate, but still.

azat commented:
>In what build configurations have you encountered them? I added suppressions to CMake files where appropriate, but still.

Ok, I see the problem, cached `HAS_SUGGEST_DESTRUCTOR_OVERRIDE` in cmake (for another compiler I guess, since I don't clean cmake cache usually)

After reset the cache for this variable everything ok.

baibaichen commented:
@azat I see the issue, it looks like we need update https://github.com/ClickHouse/ClickHouse/blob/master/docs/en/development/build-osx.md to reflect this fact.

azat commented:
This PR also removes `limit_jobs.cmake` usage, @abyss7 was it done on purpose? (if so, why `limit_jobs.cmake` had been kept?)

melin commented:
Analysis sql failure：

create table etl_test_csv
    ENGINE = MergeTree()
    ORDER BY name
    AS
    SELECT * FROM dd1_import_csv

[image: image.png]

Azat Khuzhin ***@***.***> 于2020年12月11日周五 上午2:31写道：

> This PR also removes limit_jobs.cmake usage, @abyss7
> <https://github.com/abyss7> was it done on purpose? (if so, why
> limit_jobs.cmake had been kept?)
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/ClickHouse/ClickHouse/pull/11298#issuecomment-742710216>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAIXXZT7TKMW3XQTWM7FRILSUEHWXANCNFSM4NOJTKSQ>
> .
>
